lib/ansible/modules/cloud/vmware/vmware_add_nic.py

```diff
@@ -0,0 +1,433 @@
+    argument_spec = vmware_argument_spec()
+    argument_spec.update(
+        dict(
+                vm_path=dict(required=True, type='str'),
+                label=dict(required=False, type='str'),
+                dvs=dict(required=False, type='bool', default=False),
+                state=dict(required=True, choices=['create', 'absent', 'update'], type='str'),
+                type=dict(required=False, type='str', default='vmxnet3', choices=['vmxnet','vmxnet2','vmxnet3','e1000','e1000e','pcnet32']),
+                network_name=dict(required=False, type='str'),
+                datacenter=dict(required=True, type='str'),
+                )
+        )
+
+    module = AnsibleModule(argument_spec=argument_spec,
+            required_if= [('state','create',['network_name']),
+                          ('state','update',['network_name','label']),
+                          ('state','absent',['label'])]
+            )
+
+    if not HAS_PYVMOMI:
+        module.fail_json(msg='pyvmomi is required for this module')
+
+    path = module.params.get('vm_path')
+    label = module.params.get('label')
+    dvs = module.params.get('dvs')
+    network_name = module.params.get('network_name')
+    datacenter = module.params.get('datacenter')
+    state = module.params.get('state')
+    nic_type = module.params.get('type')
+
+    changed = False
+
+    conn = connect_to_api(module)
+    proper_vm = get_vm_object(module, conn, path, datacenter)
+    all_nics = get_nics(proper_vm)
+
+    nic_type_map = dict(
+        vmxnet3 = 'VirtualVmxnet3',
+        vmxnet2 = 'VirtualVmxnet2',
+        vmxnet  = 'VirtualVmxnet',
+        e1000   = 'VirtualE1000',
+        e1000e  = 'VirtualE1000e',
+        pcnet32 = 'VirtualPCNet32'
+        )
+
+    desired_nic = dict(
+        network=network_name,
+        type=nic_type_map[nic_type],
+        dvs=dvs,
+        label=label
+        )
+
+    if state == 'create':
+        changed = True
+        pre_nics = [nic['nic_obj'] for nic in all_nics]
+        results = create_nic(module, conn, proper_vm, desired_nic)
+        new_nics_list = [nic['nic_obj'] for nic in get_nics(proper_vm)]
+        new_nic = set(new_nics_list).difference(pre_nics)
+        nic = new_nic.pop()
+        module.exit_json(changed=changed, nic=dict(label=nic.deviceInfo.label,
+                                                   key=nic.key,
+                                                   network=nic.backing.deviceName
+                                                   )
+        )
+
+    elif state == 'absent':
+        changed = True
+        results = remove_nic(module, conn, proper_vm, desired_nic, all_nics)
+
+        module.exit_json(changed=changed, msg="NIC removed")
+
+    elif state == 'update':
+        if needs_update(module, desired_nic, all_nics):
+            changed=True
+            update_nic(module, conn, proper_vm, desired_nic, all_nics)
+        else:
+            changed=False
+
+        module.exit_json(changed=changed)
+
+
```

Comment:
Could you please add integration tests for this module ? Please take a look at various VMware related tests in `test/integration` directory.